src/components/Avatar/Avatar.tsx

```diff
@@ -28,89 +29,138 @@ export interface Props {
   initials?: string;
   /** Whether the avatar is for a customer */
   customer?: boolean;
-  /** URL of the avatar image */
+  /** URL of the avatar image which falls back to initials if the image fails to load */
   source?: string;
   /** Accessible label for the avatar image */
   accessibilityLabel?: string;
 }
 
+export interface State {
+  hasError: boolean;
+  hasLoaded: boolean;
+  hasMounted: boolean;
+}
+
 export type CombinedProps = Props & WithAppProviderProps;
 
-function Avatar({
-  name,
-  source,
-  initials,
-  customer,
-  size = 'medium',
-  accessibilityLabel,
-  polaris: {intl},
-}: CombinedProps) {
-  const nameString = name || initials;
-
-  let finalSource: string | undefined;
-  let label: string | undefined;
-
-  if (accessibilityLabel) {
-    label = accessibilityLabel;
-  } else if (name) {
-    label = name;
-  } else if (initials) {
-    const splitInitials = initials.split('').join(' ');
-    label = intl.translate('Polaris.Avatar.labelWithInitials', {
-      initials: splitInitials,
-    });
-  } else {
-    label = intl.translate('Polaris.Avatar.label');
+export class Avatar extends React.PureComponent<CombinedProps, State> {
+  state: State = {
+    hasError: false,
+    hasLoaded: false,
+    hasMounted: false,
+  };
+
+  componentDidMount() {
+    this.setState({hasMounted: true});
+  }
+
+  componentDidUpdate({source: previousSource}: Props) {
+    const {source} = this.props;
+    const {hasError, hasLoaded} = this.state;
+    if (previousSource === source || (!hasError && !hasLoaded)) {
+      return;
+    }
+    // eslint-disable-next-line react/no-did-update-set-state
+    this.setState({hasError: false, hasLoaded: false});
```

Comment:
Couldn't this cause needless re-renders?

Reply:
Also appears to be untested

Reply:
I believe if state doesn't change it doesn't re-render.

We don't test the state per se but we do test it renders what it should.